### PR TITLE
Convert ShardOperationFailedException to Writeable

### PR DESCRIPTION
**Pull request by rjernst**
This commit converts subclasses of ShardOperationFailedException to
implement ctors with StreamInput instead of readFrom. It also simplifies
IndicesShardStoresResponse.Failure to serialize its shardId after the
super data.

relates #34389

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardFailure.java [hunk 1]
```diff
@@ -45,8 +45,13 @@
...
 
-    private SnapshotShardFailure() {
-
+    private SnapshotShardFailure(StreamInput in) throws IOException {
+        nodeId = in.readOptionalString();
+        shardId = new ShardId(in);
+        super.shardId = shardId.getId();
+        index = shardId.getIndexName();
+        reason = in.readString();
+        status = RestStatus.readFrom(in);
```
Comment (talevy) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotShardFailure.java]: guess this class could be a follow-up

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardFailure.java [hunk 2]
```diff
@@ -91,19 +96,7 @@ public String nodeId() {
      * @return shard failure information
      */
     static SnapshotShardFailure readSnapshotShardFailure(StreamInput in) throws IOException {
```
Comment (talevy) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotShardFailure.java]: while we're here... can we remove this, and its usage in SnapshotInfo? 😄

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardFailure.java (hunk 1 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotShardFailure.java]: This is just a static method for reading the enum. It's not from Streamable.

**Comment by rjernst** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotShardFailure.java (hunk 1 above)
Comment (talevy) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotShardFailure.java]: sure, I guess I just meant it from a design pattern perspective. anyways, let's ignore this :)

File: server/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardStoresResponse.java [hunk 3]
```diff
@@ -232,29 +233,33 @@ public Failure(String nodeId, String index, int shardId, Throwable reason) {
...
+            if (in.getVersion().onOrAfter(Version.V_7_4_0)) {
+                nodeId = in.readString();
+            }
         }
 
         public String nodeId() {
             return nodeId;
         }
 
         static Failure readFailure(StreamInput in) throws IOException {
```
Comment (talevy) [inline server/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardStoresResponse.java]: I missed this, but this seems unnecessary too

File: server/src/main/java/org/elasticsearch/action/admin/indices/close/CloseIndexResponse.java [hunk 4]
```diff
@@ -280,9 +276,7 @@ public String toString() {
             }
 
             static Failure readFailure(final StreamInput in) throws IOException {
```
Comment (talevy) [inline server/src/main/java/org/elasticsearch/action/admin/indices/close/CloseIndexResponse.java]: I missed this, but this seems unnecessary too

File: server/src/main/java/org/elasticsearch/action/support/DefaultShardOperationFailedException.java [hunk 5]
```diff
@@ -64,17 +67,14 @@ public DefaultShardOperationFailedException(String index, int shardId, Throwable
     }
 
     public static DefaultShardOperationFailedException readShardOperationFailed(StreamInput in) throws IOException {
```
Comment (talevy) [inline server/src/main/java/org/elasticsearch/action/support/DefaultShardOperationFailedException.java]: I missed this, but this seems unnecessary too

**Review by talevy — APPROVED** [review]
LGTM, only left some comments for completeness. Since these static methods are not a part of any interfaces we're trying to remove, then I'll leave it to you to decide whether to do further changes on this PR or leave it for the future.

File: server/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardStoresResponse.java (hunk 3 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardStoresResponse.java]: Given we have these all over, and haven't removed them in other PRs, I'm going to leave these for now.

**Pull request merged**